### PR TITLE
Vuejs migration

### PR DESCRIPTION
**Pull request by shubhi1092**
- Added a generic backbone view 'VueComponentView.js' which will render all the Vue Components. It is similar to RootComponent but the options in the constructor will come from the Router.
- Modified Router.coffee to use this generic backbone view to render the vue components.
- Added 2 base templates (base-empty and base-flat-empty) to use for pages without header and footer.

This will allow to directly use the vue component in the router without the need to create a backbone view with each Vue component. Eventually, we should stop using RootComponent in favor of this setup.

~~TODO~~: Documentation for how to create new pages using VueJS given this routing setup. Meanwhile, I have added an example of creating a vue component like this in this PR (VueComponentExample.vue) - will remove it before merging.

To test the example vue component, navigate to 'localhost:3000/VueComponentExample/1'

### CODE REVIEW

File: app/views/core/VueComponentView.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import store from 'core/store'
+
+const silentStore = { commit: _.noop, dispatch: _.noop }
+
+module.exports = class VueComponentView extends RootView {
+  constructor (component, options) {
+    super(options)
+    const baseTemplate = options.baseTemplate || 'base-flat'  //base template, by default using base-flat
+    this.id = 'vue-component-view'
+    this.template = require('templates/'+ baseTemplate)
```
Comment (AndrewJakubowicz) [inline app/views/core/VueComponentView.js]: Could we do an additional check prior to this ensuring that only valid base templates can be provided?

File: app/views/VueComponentExample.vue [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    id: {
+      type: Number
+    }
+  },
+  components: {
+    'pie-chart': PieChart   //using another vueComponent inside this
+  },
+  beforeCreate() {
+    // Register a vuex store module for this page, which will be destroyed when navigating away from the page
+    this.$store.registerModule('page', {
```
Comment (AndrewJakubowicz) [inline app/views/VueComponentExample.vue]: This is brilliant!

File: app/views/core/VueComponentView.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    if (this.vueComponent) {
+      this.$el.find('#site-content-area').replaceWith(this.vueComponent.$el)
+    }
+    else{
+      this.vueComponent = new this.VueComponent({
+        el: this.$el.find('#site-content-area')[0],
+        propsData: this.propsData,
+        store: store
+      })
+      window.rootComponent = this.vueComponent
```
Comment (AndrewJakubowicz) [inline app/views/core/VueComponentView.js]: Would we still need this for future components? I wonder if this example works with [Chrome Vue debugger tools](https://github.com/vuejs/vue-devtools)?

**Comment by shubhi1092** [comment]
Thank you for the review. 
`VueComponentView` is similar to RootComponent, but is able to accept arguments for propsData, base template etc so that we can use it directly in the router.
We want to deprecate RootComponent in favor of this going forward, as we do not want to create a new backbone view with every Vue component going forward.

File: app/views/core/VueComponentView.js (hunk 3 above)
Comment (shubhi1092) [reply app/views/core/VueComponentView.js]: Why do you think we would not need this?

Comment (AndrewJakubowicz) [reply app/views/core/VueComponentView.js]: We can keep this. I was wondering how much we could leverage existing Vue debugging options with this as well.

File: app/views/core/VueComponentView.js (hunk 1 above)
Comment (shubhi1092) [reply app/views/core/VueComponentView.js]: Thats a good idea, will move it to a try-catch block to catch any error in importing the base template.

File: app/views/core/VueComponentView.js (hunk 3 above)
Comment (shubhi1092) [reply app/views/core/VueComponentView.js]: We can still debug it using the chrome vue debugger tools.

**Comment by shubhi1092** [comment]
Update: If we want to use a vuex store module for a vue component page, the vue component should register it with a unique name instead of using the generic 'page' module and then unregister it before navigating away(when no longer needed).
Example here: https://github.com/codecombat/codecombat/blob/74d698a086b99facc7350f2651f5e205427585fa/app/views/VueComponentExample.vue
Will also add it in the documentation.

**Pull request merged**